### PR TITLE
Add method to check if NumberNode is set to zero

### PR DESCRIPTION
**Pull request by mtdowling**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java [hunk 1]
```diff
@@ -94,6 +95,38 @@ public Optional<NumberNode> asNumberNode() {
...
+        // Do a cheap test based on the serialized value of the number first.
+        if (toString().equals("0")) {
+            return true;
+        } else if (value instanceof BigDecimal) {
+            return value.equals(BigDecimal.ZERO);
+        } else if (value instanceof BigInteger) {
+            return value.equals(BigInteger.ZERO);
+        } else if (value instanceof Float) {
+            // Using Math.signum to avoid epsilon issues.
+            return Math.signum((Float) value) == 0;
```
Comment (gosar) [inline smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java:115]: I'm unsure if we want isZero to return true of false for negative zero. From signum javadoc:
>Special Cases:
    If the argument is NaN, then the result is NaN.
    If the argument is positive zero or negative zero, then the result is the same as the argument. 

So if we want isZero false, this seems fine. But if we want isZero true, then we may need to add another `||` condition here for negative zero.

Either ways, we should add javadoc comment and test case for negative zero.

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java [hunk 2]
```diff
@@ -94,6 +95,38 @@ public Optional<NumberNode> asNumberNode() {
...
+    public boolean isZero() {
+        // Do a cheap test based on the serialized value of the number first.
+        if (toString().equals("0")) {
+            return true;
+        } else if (value instanceof BigDecimal) {
+            return value.equals(BigDecimal.ZERO);
+        } else if (value instanceof BigInteger) {
+            return value.equals(BigInteger.ZERO);
+        } else if (value instanceof Float) {
+            // Using Math.signum to avoid epsilon issues.
```
Comment (gosar) [inline smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java:114]: If there is a test case that shows the epsilon issue that would have broken without this behavior, that'd be great.

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java (hunk 1 above)
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java:115]: I think -0 and +0 are fine. I'll add tests.

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java [hunk 3]
```diff
@@ -94,6 +95,38 @@ public Optional<NumberNode> asNumberNode() {
...
+            return true;
+        } else if (value instanceof BigDecimal) {
+            return value.equals(BigDecimal.ZERO);
+        } else if (value instanceof BigInteger) {
+            return value.equals(BigInteger.ZERO);
+        } else if (value instanceof Float) {
+            return value.floatValue() == 0;
+        } else {
+            return value.doubleValue() == 0;
+        }
```
Comment (gosar) [inline smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java:127]:
```suggestion
        } else if (value instanceof Float) {
            return value.floatValue() == 0f;
        } else {
            return value.doubleValue() == 0d;
        }
```

**Review by gosar — COMMENTED** [review]
ship it after this last change requested

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java (hunk 3 above)
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java]: updated

**Review by gosar — APPROVED** [review]
